### ISSUE TITLE
Removing backward compatibility kludge from -zip

### ISSUE THREAD

**Issue by tarsius** [magnars/dash.el#135]
For backward compatibility reasons `-zip` return a cons-cell instead of a list with two elements when called on two lists. Unfortunately this means that doing `(apply #'-zip list-of-lists)` is risky when the length of `list-of-lists` is unknown.

Could you please consider removing this kludge? Alternatively, if you would rather avoid that, maybe you could add `-zip-apply (lists)`. But I think ripping of the bandage would be better.

The kludge was added in #82.

**Comment by yangchenyun** [comment]
I was byte by this issue in my code.

`-zip' returns value is inconsistent in this case, for list of two, each element is `(cons n m)', for list of three, each element is `(cons n (cons m (cons p '()))' 

The terminal of recursion is different and makes the result list inconsistent.

**Comment by magnars** [comment]
Ack, that is terrible. Yes, this backwards compatibility kludge must die.

**Comment by Fuco1** [comment]
Is it a kludge though? Zip originally supported only two lists and the multivariant version whas added later.  Simply a bad design decision on my part (I think I implemented the original zip).

I suppose not many people rely on it, but if we're going to put out a BC-breaking release, let's fix all the crap at once?

**Comment by yangchenyun** [comment]
I would vote for:
- an alias for the old API for zip two lists, `-zip-pair`(?), and encourage people to migrate right now if they rely on the old behavior. 
- a BC-breaking release for the shiny and consistent behavior of `-zip` over multiple lists

**Comment by magnars** [comment]
@yangchenyun :+1:

**Comment by tarsius** [comment]
Yes, please do that.

**Comment by magnars** [comment]
I have now added the `-zip-pair` alias, and added a warning to the docstring and readme, in preparation for the upcoming breaking changes in 2.0 dash.

**Comment by tarsius** [comment]
Thanks!

**Comment by Fuco1** [comment]
@magnars any time frame? I have some last things I'd like to polish, plus we should go over the old naming issue and maybe fix all of that at once for the BC release?

**Comment by magnars** [comment]
@Fuco1 I think you know better what should go into 3.0. I hope this will be the last release with major breaking changes in a long while, so yeah, certainly include the things we want done. Maybe we should try listing them as issues and marking them with a milestone label?

**Comment by Fuco1** [comment]
@magnars All right, I'll set up the issues over the weekend, is it OK or do you want it sooner?

**Comment by magnars** [comment]
There's no rush.
ons. 21. okt. 2015 kl. 18.25 skrev Matus Goljer <notifications@github.com>:

**Comment by Wilfred** [comment]
I'd really like to see this land in dash.el. In the meantime, is there any alternative function for explicitly using proper lists? I can't see one.

**Comment by Fuco1** [comment]
You would have to map over it now.

**Comment by Wilfred** [comment]
Would there be any interest in a `-zip-lists` function that provided this functionality today?

**Comment by tarsius** [comment]
Uhm. How is 3.0.0 coming along?

**Comment by tarsius** [comment]
It has been three years. Maybe it is time to do this now?
Anyway, I am in the process of closing my old issues and will do so for this one too eventually.

**Comment by Wilfred** [comment]
There's no good way of breaking backwards compatibility in elisp. I'm not sure it's worth changing, although I think a `-zip-lists` would be reasonable.

That being said, I can't see many packages that depend on `-zip`. There are only two in my current Emacs instance, according to elisp-refs.

**Comment by magnars** [comment]
I agree with Wilfred. Breaking changes are broken. Let's use a new name instead.

**Comment by tarsius** [comment]
That works for me too.

**Comment by Fuco1** [comment]
So the solution is:

* we keep `-zip` working as it does.
* there is an alias `-zip-pair` for `-zip` which we can keep but is pretty much redundant
* we introduce `-zip-lists` which will always return a list, even for 0, 1, 2... arguments.

I've looked through all my projects and personally I only used `-zip` with 2 arguments 6 times.  I think if we "bite the bullet" and do a BC break not many people would be affected.

We could also try to search github for `-zip` usage and offer PRs to the project, wait a month and do it.

**Comment by Fuco1** [comment]
Ah well, there's over 500 results for `-zip` on github... so probably not a viable strategy.

**Comment by khinsen** [comment]
I found this discussion after wondering about a sentence in the README: "If you rely on this behavior, use -zip-pair instead." What exactly does "relying on" mean? More concretely, how should I write my code if I want it to work before and after the change? The only way I see not to "rely" on the current behavior is not to use `-zip` at all when the number of lists might be two. I'd then go for `-zip-pair` when I know I have two lists and I write my own work-around (probably based on `-zip-with`) when the number of lists is variable.

In other words, I'd appreciate more concrete guidance in the README for preparing for the change!

**Comment by Fuco1** [comment]
If you want to future proof your code and you know you have two lists and you want a dotted pair out, use `-zip-pair`.

If you have more than two lists, use `-zip` and there won't be any change.

If you use `-zip` on two lists the code will break after the update.

If you are doing something like `(apply '-zip lists)` and you don't know what `lists` is you will need to guard this with `(length lists)` and some magic to make the code future-proof.

This is not ideal.

I think adding `-zip-lists` or something which always returns lists makes sense.  We can later alias `zip` to that one although maybe it's not worth the fuss.

One day I will just do it and handle all the abuse for a couple days but then it will be over us

**Comment by magnars** [comment]
No breaking changes please. A correct version with a new name is 👍.
ons. 31. okt. 2018 kl. 21:47 skrev Matus Goljer <notifications@github.com>:

**Comment by tarsius** [comment]
Friendly ping. :beer:

**Comment by Fuco1** [comment]
I'm doing this right now.  New version will be tagged this evening!

**Comment by Fuco1** [comment]
Function added in 52fd010, tagged new 2.17.0 release as well :tada:

**Comment by tarsius** [comment]
:clinking_glasses: Thanks!

We are keeping `-zip`, not just `-zip-pair`, and can close this issue, right?

**Comment by khinsen** [comment]
Thanks for implementing a workable solution!

**Comment by Fuco1** [comment]
@tarsius `-zip` will keep working as it was, `-zip-pair` is an alias to `-zip` and `-zip-lists` is a new function which always returns list of lists, with any number of input arguments.
